server/src/main/java/io/druid/server/SegmentManager.java

```diff
@@ -93,35 +136,33 @@ public boolean loadSegment(final DataSegment segment) throws SegmentLoadingExcep
   {
     final Segment adapter = getAdapter(segment);
 
-    synchronized (lock) {
-      final String dataSource = segment.getDataSource();
-      final VersionedIntervalTimeline<String, ReferenceCountingSegment> loadedIntervals = dataSources.computeIfAbsent(
-          dataSource,
-          k -> new VersionedIntervalTimeline<>(Ordering.natural())
-      );
-
-      final PartitionHolder<ReferenceCountingSegment> entry = loadedIntervals.findEntry(
-          segment.getInterval(),
-          segment.getVersion()
-      );
-      if ((entry != null) && (entry.getChunk(segment.getShardSpec().getPartitionNum()) != null)) {
-        log.warn("Told to load a adapter for a segment[%s] that already exists", segment.getIdentifier());
-        return false;
-      }
+    final DataSourceState computeResult = dataSources.compute(
```

Comment:
Please leave a comment, that it uses compute to ensure operations isolation with per-dataSource granularity.